### DISCUSSION TITLE
API Versioning

### DISCUSSION THREAD

**Discussion by andrew-hoskin** [MEF-GIT/MEF-LSO-Sonata-SDK#188] [General]
The MEF documentation, provides the following url model: 
Quote - 'https://{serverBase}/mefApi/sonata/quoteManagement/v8/'.

According to this, we should only change the {serverBase} part.

At Connectbase, we are looking at supporting multiple SDK versions (Billie, Dolly etc), but, each API of the SDK versions has its own version;

For example, looking at the Quote API: 
- Bille → 7.0.0-RC; 
- Celine → 8.0.0; 
- Dolly → 8.0.1; 
- Ella → 8.0.1; 
- Fergie → 8.0.1; 
- Grace → 8.0.1).

For the Quote API, in this case we would have the following endpoints : 

- Billie  '.../quoteManagement/v7'; 
- Celine - '.../quoteManagement/v8'; 
- Dolly - '.../quoteManagement/v8', ....

The issue is that Celine and Dolly have the same '/v8' and we can't differentiate them.

**Options:**
1.) In the {serverBase}, we can add the sdk version i.e. '/dolly' -> 'https://api.connectbase.com/dolly/mefApi/sonata/quoteManagement/v8/', but then if you look at Dolly, Ella, Fergie and Grace
they have the same version for Quote API (8.0.1), so we'll need a different endpoint for each (i.e.'/ella'...) which will point to the same code.

Depending on the versions we use, we'll end up with many endpoints that point to the same code.

2.) Use complete API version: instead of '/v8' - use '/v800'(Celine) and '/v801' (Dolly, Ella, Fergie, Grace). In this case we won't have so many endpoints, but it doesn't use the same url model as defined by MEF. And we would need to know exactly that v800 is Celine, v801 is Dolly...
'https://api.connectbase.com/mefApi/sonata/quoteManagement/v801/'

I suspect this may not be in line with the MEF standard.

Has anyone else encountered this issue and is there a suggested or recommended approach?

Many thanks,

Andrew

**Comment by fahimsabir-colt** [comment]
There isn’t a suggested approach because most implementations have only one version in production, and thus won’t have hit this issue.

At Colt, we regard the API format that is suggested as not being a mandatory element of the specification (everyone has their own URL structuring schemes, and it would be wrong of MEF to force one onto carriers) and have chosen to go a different route.


  *   We use the SDK version name in the URL, because it is how (most) people reference it.
  *   We have removed the interface because the APIs are the same for both Sonata and Cantata
  *   We have removed the version number because the versions are just confusing
  *   We have removed the API family as it just creates unnecessary noise.

So, rather than:
https://ondemand.colt.net/mefapi/sonata/quotemanagement/v7/quote

We have used:
https://mefapi.ondemand.colt.net/billie/quote


From: Andrew ***@***.***>
Sent: 25 April 2024 15:55
To: MEF-GIT/MEF-LSO-Sonata-SDK ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [MEF-GIT/MEF-LSO-Sonata-SDK] API Versioning (Discussion #188)

EXTERNAL EMAIL*

The MEF documentation, provides the following url model:
Quote - 'https://{serverBase}/mefApi/sonata/quoteManagement/v8/'.

According to this, we should only change the {serverBase} part.

At Connectbase, we are looking at supporting multiple SDK versions (Billie, Dolly etc), but, each API of the SDK versions has its own version;

For example, looking at the Quote API:

  *   Bille → 7.0.0-RC;
  *   Celine → 8.0.0;
  *   Dolly → 8.0.1;
  *   Ella → 8.0.1;
  *   Fergie → 8.0.1;
  *   Grace → 8.0.1).

For the Quote API, in this case we would have the following endpoints :

  *   Billie '.../quoteManagement/v7';
  *   Celine - '.../quoteManagement/v8';
  *   Dolly - '.../quoteManagement/v8', ....

The issue is that Celine and Dolly have the same '/v8' and we can't differentiate them.

Options:
1.) In the {serverBase}, we can add the sdk version i.e. '/dolly' -> 'https://api.connectbase.com/dolly/mefApi/sonata/quoteManagement/v8/', but then if you look at Dolly, Ella, Fergie and Grace
they have the same version for Quote API (8.0.1), so we'll need a different endpoint for each (i.e.'/ella'...) which will point to the same code.

Depending on the versions we use, we'll end up with many endpoints that point to the same code.

2.) Use complete API version: instead of '/v8' - use '/v800'(Celine) and '/v801' (Dolly, Ella, Fergie, Grace). In this case we won't have so many endpoints, but it doesn't use the same url model as defined by MEF. And we would need to know exactly that v800 is Celine, v801 is Dolly...
'https://api.connectbase.com/mefApi/sonata/quoteManagement/v801/'

I suspect this may not be in line with the MEF standard.

Has anyone else encountered this issue and is there a suggested or recommended approach?

Many thanks,

Andrew

—
Reply to this email directly, view it on GitHub<https://github.com/MEF-GIT/MEF-LSO-Sonata-SDK/discussions/188>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AASM5DFFXA57RDQWKNUXBBDY7EKLTAVCNFSM6AAAAABGZB5WSOVHI2DSMVQWIX3LMV43ERDJONRXK43TNFXW4OZWGU2TSOJQG4>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.******@***.***>>
* WARNING - EXTERNAL: This email originated from outside of Colt. Do not click any links or open any attachments unless you trust the sender and know that the content is safe.

[Colt Disclaimer] This email is from an entity of the Colt group of companies. Colt Group Holdings Limited, Colt House, 20 Great Eastern Street, London, EC2A 3EH, United Kingdom, registered in England and Wales, under company number 11530966. Corporate and contact information for our entities can be found at https://www.colt.net/legal/colt-group-of-companies/. Internet communications are not secure and Colt does not accept responsibility for the accurate transmission of this message. Content of this email or its attachments is not legally or contractually binding unless expressly previously agreed in writing by Colt

**Reply by MichalLaczynskiAmartus** [reply]
Thanks @fahimsabir-colt for stepping in :)

**Comment by MichalLaczynskiAmartus** [comment]
Hi @andrew-hoskin 

As Fahim said, the final structure of the URL is open.
From the detailed implementation that I am aware of - in Amartus we use your suggested approach no1:
- it keeps the remainder of the urn as suggested in the standard
- allows differentiation by SDK release, which is very important, because while the version of Quote did not change e.g. from Dolly to Ella, the Product Order did. And thus the code receiving a request in SDK Ella needs to work with Ella Product Order (and so Dolly -> Dolly), which is not possible to recognize when you're using only the single API version.

**Comment by andrew-hoskin** [comment]
Thanks both for the fast reply. I like the simplicity of your approach Fathom. I will review with the team and come back with any further questions.
